### PR TITLE
#2325 add namespace support for policies

### LINKED ISSUE #2325: Add optional `namespaces` field to policy entries for namespace-scoped authorization
## Summary

Add an optional `namespaces` field to policy entries that restricts the entry's applicability to things within matching namespaces. This enables namespace-scoped authorization without changes to the search index structure.

## Motivation

Policy entries currently apply universally regardless of the thing's namespace. This limitation affects:

1. **Multi-Tenant Scenarios**: Organizations need tenant isolation without maintaining separate policies per tenant
2. **Namespace-Based Organization**: Different user groups need different permissions per namespace (e.g., vehicle managers vs. facility managers)

## Proposed Solution

Add an optional `namespaces` array to policy entries:

```json
{
  "policyId": "com.acme:shared-policy",
  "entries": {
    "vehicle-admins": {
      "subjects": { "nginx:fleet-manager": { "type": "generated" } },
      "resources": { "thing:/": { "grant": ["READ", "WRITE"], "revoke": [] } },
      "namespaces": ["com.acme.vehicles", "com.acme.vehicles.*"]
    },
    "analysts": {
      "subjects": { "nginx:data-analyst": { "type": "generated" } },
      "resources": { "thing:/": { "grant": ["READ"], "revoke": [] } }
    }
  }
}
```

### Semantics

| `namespaces` field | Behavior |
|--------------------|----------|
| Absent | Entry applies to **all namespaces** (backward compatible) |
| Empty array `[]` | Same as absent |
| Array with patterns | Entry **only applies** when thing's namespace matches at least one pattern |

### Namespace Pattern Matching

- `com.acme` - exact match only
- `com.acme.*` - matches everything below `com.acme` but **not** `com.acme` itself

To match both: `["com.acme", "com.acme.*"]`

## Multi-Tenant Example

```json
{
  "policyId": "platform:multi-tenant-base",
  "entries": {
    "tenant-a-users": {
      "subjects": { "oidc:tenant-a-group": {} },
      "resources": { "thing:/": { "grant": ["READ", "WRITE"], "revoke": [] } },
      "namespaces": ["com.tenant-a", "com.tenant-a.*"]
    },
    "tenant-b-users": {
      "subjects": { "oidc:tenant-b-group": {} },
      "resources": { "thing:/": { "grant": ["READ", "WRITE"], "revoke": [] } },
      "namespaces": ["com.tenant-b", "com.tenant-b.*"]
    }
  }
}
```

Tenant A users **cannot** access Tenant B things (and vice versa).

## Implementation Notes

- **Search index unchanged**: Namespace filtering applied at index write time in `EvaluatedPolicy.of()` - entries filtered by thing's namespace before extracting subjects
- **Backward compatible**: Absent field means "applies to all namespaces"
- **Validation**: Reuse existing `RegexPatterns.NAMESPACE_PATTERN` with `*` allowed as trailing wildcard

## Related

- #2304 - Complementary feature for gateway-level namespace enforcement based on JWT claims
- #2347 - "entriesAdditions" - will be enhanced to also be able to configure additional "namespaces" as part of this issue

### PR DESCRIPTION
**Pull request by hu-ahmed**
Resolves #2325

Add support for namespace-scoped policy entries via a new optional `namespaces` field on `PolicyEntry`.

A scoped policy entry only applies to Things whose namespace matches at least one configured pattern.
If `namespaces` is omitted or empty, the entry remains globally applicable for backward compatibility.

Supported matching semantics:
- `com.acme` matches only the exact namespace `com.acme`
- `com.acme.*` matches nested namespaces such as `com.acme.vehicles`, but not `com.acme` itself

### CODE REVIEW

File: policies/enforcement/src/main/java/org/eclipse/ditto/policies/enforcement/PolicyEnforcer.java [hunk 1]
```diff
@@ -110,4 +114,28 @@ public Enforcer getEnforcer() {
...
+    public PolicyEnforcer forNamespace(final String thingNamespace) {
+        if (policy == null) {
+            return this;
+        }
+        final boolean anyEntryHasNamespaces = StreamSupport.stream(policy.spliterator(), false)
+                .anyMatch(entry -> !entry.getNamespaces().isEmpty());
+        if (!anyEntryHasNamespaces) {
+            return this;
+        }
+        final List<PolicyEntry> filteredEntries = StreamSupport.stream(policy.spliterator(), false)
```
Comment (thjaeckle) [inline policies/enforcement/src/main/java/org/eclipse/ditto/policies/enforcement/PolicyEnforcer.java]: ### MEDIUM: Double iteration over policy entries

The code iterates through all policy entries twice:
1. Lines 129-130: `anyMatch(entry -> !entry.getNamespaces().isEmpty())` — full scan to check if any entry has namespaces
2. Lines 134-136: `filter(entry -> entry.appliesToNamespace(...))` — full scan again to filter

This can be merged into a single pass:
```java
final List<PolicyEntry> filteredEntries = StreamSupport.stream(policy.spliterator(), false)
        .filter(entry -> entry.appliesToNamespace(thingNamespace))
        .collect(Collectors.toList());
if (filteredEntries.size() == (int) StreamSupport.stream(policy.spliterator(), false).count()) {
    return this; // no entries were filtered out
}
```

Or more efficiently, track whether any entry was actually filtered:
```java
boolean anyFiltered = false;
final List<PolicyEntry> filtered = new ArrayList<>();
for (final PolicyEntry e : policy) {
    if (e.appliesToNamespace(thingNamespace)) {
        filtered.add(e);
    } else {
        anyFiltered = true;
    }
}
if (!anyFiltered) return this;
```

Note: `forNamespace()` is called on every enforcement via `ThingEnforcerActor.providePolicyEnforcer()`, so this runs on every thing command.

File: policies/model/src/main/java/org/eclipse/ditto/policies/model/PolicyEntry.java [hunk 2]
```diff
@@ -111,6 +112,48 @@ default JsonSchemaVersion[] getSupportedSchemaVersions() {
...
+     * @param namespace the thing's namespace to check against.
+     * @return {@code true} if this entry applies to the given namespace.
+     * @since 3.9.0
+     */
+    default boolean appliesToNamespace(final String namespace) {
+        final List<String> patterns = getNamespaces();
+        if (patterns.isEmpty()) {
+            return true;
+        }
+        for (final String pattern : patterns) {
```
Comment (thjaeckle) [inline policies/model/src/main/java/org/eclipse/ditto/policies/model/PolicyEntry.java]: ### MEDIUM: Inconsistency with `NamespacePoliciesConfig.namespaceMatchesPattern()` from PR #2367

This method supports two pattern types: exact match and `prefix.*` wildcard.
PR #2367's `NamespacePoliciesConfig.namespaceMatchesPattern()` supports the same two PLUS `*` (catch-all).

While the validator in `ImmutablePolicyEntry.isValidNamespacePattern()` correctly rejects `*` for policy entries, the two matching methods have divergent implementations:
- Here: `namespace.startsWith(prefix + ".")` (concatenates `"."` each call)
- PR #2367: `namespace.startsWith(pattern.substring(0, pattern.length() - 1))` (the substring already includes the dot)

Consider extracting a shared static utility method to avoid subtle divergence as the codebase evolves. Both PRs introduce namespace pattern matching independently — having two slightly different implementations is a maintenance risk.

File: policies/enforcement/src/test/java/org/eclipse/ditto/policies/enforcement/PolicyEnforcerTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                Collections.singletonList(Subject.newInstance(SubjectId.newInstance(subjectId))),
+                Collections.singletonList(PoliciesModelFactory.newResource("thing", "/",
+                        EffectedPermissions.newInstance(
+                                PoliciesModelFactory.newPermissions(Permission.READ),
+                                PoliciesModelFactory.noPermissions()))),
+                ImportableType.IMPLICIT,
+                Collections.emptySet(),
+                namespaces);
+    }
+}
```
Comment (thjaeckle) [inline policies/enforcement/src/test/java/org/eclipse/ditto/policies/enforcement/PolicyEnforcerTest.java:105]: ### MEDIUM: Missing test for `forNamespace()` edge case

The production code handles `policy == null` at `PolicyEnforcer.java:126`:
```java
if (policy == null) {
    return this;
}
```

But no test covers this path. Consider adding:
```java
@Test
public void forNamespaceWithNullPolicyReturnsSameInstance() {
    final PolicyEnforcer enforcer = new PolicyEnforcer(null,
            PolicyEnforcers.defaultEvaluator(Collections.emptyList()));
    assertThat(enforcer.forNamespace("any.ns")).isSameAs(enforcer);
}
```

Also: `forNamespace()` doesn't guard against `thingNamespace == null`. If a thing has no namespace set (shouldn't happen but defensively), `startsWith(prefix + ".")` would NPE. Consider adding a null check or documenting the precondition.

File: policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyEntry.java [hunk 4]
```diff
@@ -176,6 +213,44 @@ private static Set<AllowedImportAddition> readAllowedImportAdditions(final JsonO
...
+            throw PolicyEntryInvalidException.newBuilder()
+                    .description("The value '" + value + "' in field '" +
+                            JsonFields.NAMESPACES.getPointer() +
+                            "' is not a valid namespace pattern. Valid patterns are namespaces" +
+                            " like 'com.acme' or wildcard patterns like 'com.acme.*'.")
+                    .build();
+        }
+    }
+
+    private static boolean isValidNamespacePattern(final String value) {
```
Comment (thjaeckle) [inline policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyEntry.java]: ### MEDIUM: No duplicate validation for namespace patterns

The list `["com.acme", "com.acme"]` passes validation since each element is individually valid. While not harmful at runtime (the duplicate just causes a redundant match check), it could indicate a user misconfiguration.

Consider either:
- Deduplicating silently (e.g., `new ArrayList<>(new LinkedHashSet<>(namespaces))`)
- Or rejecting duplicates with a descriptive error

Also: the list preserves insertion order, which is good for deterministic behavior. But consider whether `Set<String>` would better represent the semantics (unordered, no duplicates). The current `List<String>` choice is fine for JSON round-tripping though.

File: documentation/src/main/resources/pages/ditto/basic-policy.md [hunk 5]
```diff
@@ -340,7 +354,8 @@ Example of a policy specifying different types of `importable` entries and allow
       "subjects": { ... },
       "resources": { ... },
       "importable": "implicit",
-      "allowedImportAdditions": [ "subjects" ]
+      "allowedImportAdditions": [ "subjects" ],
+      "namespaces": [ "com.acme", "com.acme.*" ]
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-policy.md]: I would move that to below `subjects` and `resources` - does not really matter from JSON perspective - but is more logically to have those 3 "together".

File: documentation/src/main/resources/openapi/sources/schemas/policies/policyEntry.yml [hunk 6]
```diff
@@ -19,6 +19,8 @@ properties:
     $ref: 'importable.yml'
   allowedImportAdditions:
     $ref: 'allowedImportAdditions.yml'
+  namespaces:
```
Comment (thjaeckle) [inline documentation/src/main/resources/openapi/sources/schemas/policies/policyEntry.yml]: I would move that up to below `resources` definition

File: documentation/src/main/resources/pages/ditto/basic-policy.md
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-policy.md]: This needs to be enhanced with `"namespaces"` as well.

File: policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyBuilder.java [hunk 7]
```diff
@@ -56,6 +58,7 @@ private ImmutablePolicyBuilder() {
         revokedPermissions = new LinkedHashMap<>();
         importableTypes = new LinkedHashMap<>();
         allowedImportAdditions = new LinkedHashMap<>();
+        namespaces = new LinkedHashMap<>();
```
Comment (thjaeckle) [inline policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyBuilder.java]: I would also move this declaration up - below `revokedPermissions`

File: policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyBuilder.java [hunk 8]
```diff
@@ -42,6 +43,7 @@ final class ImmutablePolicyBuilder implements PolicyBuilder {
     private final Map<Label, Map<ResourceKey, Permissions>> revokedPermissions;
     private final Map<Label, ImportableType> importableTypes;
     private final Map<Label, Set<AllowedImportAddition>> allowedImportAdditions;
+    private final Map<Label, List<String>> namespaces;
```
Comment (thjaeckle) [inline policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyBuilder.java]: please move this declaration up - above `importableTypes`

File: policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyEntry.java [hunk 9]
```diff
@@ -111,7 +118,35 @@ public static PolicyEntry of(final Label label, final Subjects subjects, final R
...
+     * @param importableType specifies whether and how this entry is allowed to be imported by others.
+     * @param allowedImportAdditions which types of additions are allowed when importing this entry.
+     * @param namespaces the namespace patterns restricting which thing namespaces this entry applies to.
+     * @return a new {@code PolicyEntry} object.
+     * @throws NullPointerException if any argument is {@code null}.
+     * @since 3.9.0
+     */
+    public static PolicyEntry of(final Label label, final Subjects subjects, final Resources resources,
+            final ImportableType importableType, final Set<AllowedImportAddition> allowedImportAdditions,
+            final List<String> namespaces) {
```
Comment (thjaeckle) [inline policies/model/src/main/java/org/eclipse/ditto/policies/model/ImmutablePolicyEntry.java]: I would not create the `namespaces` parameter as last one - but move behind `resources`

**Review by thjaeckle — APPROVED** [review]
@hu-ahmed PR is in a great shape, thanks a lot 👍 

Did a final review - and added as part of that to not by default fallback to `"namespaces": []` if none were explicitly present/persisted (e.g. when querying the new `GET /.../namespaces` this was returned for policies without any namespaces defined).
